File: althex/from_althex.py
def from_althex(althex_string, use_uppercase=None):
    """
    Converts an althex string, where 'G'-'Z' or 'g'-'z' represent '0'-'9' and 'A'-'F', back to the standard hexadecimal format.
    
    The case for the output can be explicitly set. If not set, it is determined by the case of the characters in the althex string.
    If `is_upper_althex` returns True for the althex string, the output will be uppercase. If `is_upper_althex` returns False, which
    includes cases with all lowercase or mixed case characters, the output will default to lowercase.
    
    Parameters:
    althex_string (str): The althex string to be converted back to standard hexadecimal format.
    use_uppercase (bool, optional): If set to True, the output is forced to uppercase. If set to False, the output is forced to lowercase.
                                    If None, the case is inferred from the althex_string by `is_upper_althex`.
    
    Returns:
    str: A standard hexadecimal string representing the same value as the input althex string. The case of the output is determined by
         the `use_uppercase` parameter or by the case of the input string if `use_uppercase` is None.
    
    Examples:
    >>> from_althex('HUKZ')
    '1A3F'
    >>> from_althex('hukz')
    '1a3f'
    >>> from_althex('HUKZ', use_uppercase=True)
    '1A3F'
    >>> from_althex('hukz', use_uppercase=False)
    '1a3f'
    >>> from_althex('HuKz')
    '1a3f'  # Mixed case without explicit 'use_uppercase' results in lowercase output.
    """
    
    # Determine the case from the input string if use_uppercase is not set
    if use_uppercase is None:
        use_uppercase = is_upper_althex(althex_string)

    # Create the translation map based on the determined case
    custom_to_hex_map = str.maketrans(
        'GHJKMNPRSTUVWXYZ' + 'ghjkmnprstuvwxyz',
        '0123456789ABCDEF' * 2 if use_uppercase else '0123456789abcdef' * 2
    )

    # Translate the custom hex string back to standard hex
    return althex_string.translate(custom_to_hex_map)

File: glitch-projects/ehime-u/althex/althex-js/from_althex.js
function fromAlthex(althexString, useUppercase = null) {
    // Helper function to determine if the string is upper althex
    function isUpperAlthex(s) {
        const althexChars = new Set('GHJKLMNOPQRSTUVWXYZ');
        const filteredChars = [...s].filter(c => althexChars.has(c));
        return filteredChars.length === s.length && filteredChars.every(c => c === c.toUpperCase());
    }

    // Determine the case from the input string if useUppercase is not set
    if (useUppercase === null) {
        useUppercase = isUpperAlthex(althexString) && althexString === althexString.toUpperCase();
    }

    // Create the translation map based on the determined case
    const althexMapUpper = { 'G': '0', 'H': '1', 'J': '2', 'K': '3', 'M': '4', 'N': '5', 'P': '6', 'R': '7', 'S': '8', 'T': '9', 'U': 'A', 'V': 'B', 'W': 'C', 'X': 'D', 'Y': 'E', 'Z': 'F' };
    const althexMapLower = { 'g': '0', 'h': '1', 'j': '2', 'k': '3', 'm': '4', 'n': '5', 'p': '6', 'r': '7', 's': '8', 't': '9', 'u': 'a', 'v': 'b', 'w': 'c', 'x': 'd', 'y': 'e', 'z': 'f' };

    // Translate the althex string back to standard hex
    return [...althexString].map(c => {
        if (useUppercase) {
            return althexMapUpper[c.toUpperCase()] || c;
        } else {
            return althexMapLower[c.toLowerCase()] || c;
        }
    }).join('');
}
